clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -5335,6 +5335,217 @@ static void handlePreferredTypeAttr(Sema &S, Decl *D, const ParsedAttr &AL) {
+
+      // If there are already missing attributes with same arguments, do not add
+      // duplicates.
+      for (FormatAttr *FA : CompoundStmtMissingAttributes) {
+        if (!llvm::any_of(MissingAttributes, [&](const FormatAttr *Attr) {
+              return FA->getType() == Attr->getType() &&
+                     FA->getFormatIdx() == Attr->getFormatIdx() &&
+                     FA->getFirstArg() == Attr->getFirstArg();
+            }))
+          MissingAttributes.push_back(FA);
+      }
+
+      continue;
+    }
+
+    ValueStmt *VS = dyn_cast_or_null<ValueStmt>(Child);
+    if (!VS)
+      continue;
+    Expr *TheExpr = VS->getExprStmt();
+    if (!TheExpr)
+      continue;
+    CallExpr *TheCall = dyn_cast_or_null<CallExpr>(TheExpr);
+    if (!TheCall)
+      continue;
+    const FunctionDecl *ChildFunction =
+        dyn_cast_or_null<FunctionDecl>(TheCall->getCalleeDecl());
+    if (!ChildFunction || !ChildFunction->hasAttr<FormatAttr>())
+      continue;
+
+    Expr **Args = TheCall->getArgs();
+    unsigned int NumArgs = TheCall->getNumArgs();
+
+    // If child expression is function, check if it is format function.
+    // If it is, check if parent function misses format attributes.
+
+    unsigned int ChildFunctionFormatArgumentIndexOffset =
+        checkIfMethodHasImplicitObjectParameter(ChildFunction) ? 2 : 1;
+
+    // If child function is format function and format arguments are not
+    // relevant to emit diagnostic, save only information about format type
+    // (format index and first-to-check argument index are set to -1).
+    // Information about format type is later used to determine if there are
+    // more than one format type found.
+
+    // Check if function has format attribute with forwarded format string.
+    IdentifierInfo *AttrType;
+    const ParmVarDecl *FormatArg;
+    if (!llvm::any_of(ChildFunction->specific_attrs<FormatAttr>(),
+                      [&](const FormatAttr *Attr) {
+                        AttrType = Attr->getType();
+
+                        int OffsetFormatIndex =
+                            Attr->getFormatIdx() -
+                            ChildFunctionFormatArgumentIndexOffset;
+                        if (OffsetFormatIndex < 0 ||
+                            (unsigned)OffsetFormatIndex >= NumArgs)
+                          return false;
+
+                        const auto *FormatArgExpr = dyn_cast<DeclRefExpr>(
+                            Args[OffsetFormatIndex]->IgnoreParenCasts());
+                        if (!FormatArgExpr)
+                          return false;
+
+                        FormatArg = dyn_cast_or_null<ParmVarDecl>(
+                            FormatArgExpr->getReferencedDeclOfCallee());
+                        if (!FormatArg)
+                          return false;
+
+                        return true;
+                      })) {
+      MissingAttributes.push_back(
+          FormatAttr::CreateImplicit(getASTContext(), AttrType, -1, -1));
```

Comment:
But one of those calls might not be forwarding the parameter, and thus be irrelevant:
```c++
void f(char *in, va_list args)
{
    printf("Accepting input: ");
    vscanf(out, args);
}
```
Didn't try it, but it looks like this might not emit the warning because of `printf`. And forwarding to both `printf` and `scanf` doesn't really make sense to me, or at least it's probably rare enough that we don't need to worry about it.

Reply:
I tested this function and it emits diagnostic due to va_list checking (snippet below). As va_list is not passed to printf, format type info is not saved, so diagnostic is emitted. If this snippet was removed, there would be no warning.

```
    // Check if va_list is passed to callee function.
    // If va_list is not passed, continue to the next iteration.
    bool hasVaList = false;
    for (const auto *Param : CalleeFunction->parameters()) {
      if (Param->getOriginalType().getCanonicalType() ==
          S.getASTContext().getBuiltinVaListType().getCanonicalType()) {
        hasVaList = true;
        break;
      }
    }
    if (!hasVaList)
      continue;
```

Reply:
Fair enough, but I think it's too complicated. Passing the `va_list` to multiple functions with different attributes should be extremely rare, and I don't see how it can make sense. So warning about it is probably fine.